File: plugins/slight_replica/tools/deploy_plugin.py
#!/usr/bin/env python3
"""Deploy slight_replica to a declared Eden or Ryujinx mod directory."""

from __future__ import annotations

import argparse
import os
import shutil
from pathlib import Path

from host_paths import (
    eden_mod_directory,
    ryujinx_mod_directory,
    ryujinx_sd_directory,
)

RUNTIME_DEPENDENCIES = (
    "libarcropolis.nro",
    "libnro_hook.nro",
    "libsmashline_plugin.nro",
)
SKYLINE_EXEFS = ("subsdk9", "main.npdm")
INSTALLED_NAME = "lib_effect_viewer.nro"
LEGACY_NAME = "libeffect_viewer.nro"

# Any file in the plugins directory whose bytes contain this is a copy of us.
#
# Skyline loads EVERY file in romfs:/skyline/plugins/ as a plugin, extension ignored. A
# `lib_effect_viewer.nro.bak` left beside the real one therefore runs a second full copy: two sets
# of ACMD hooks, two per-frame drivers, two servers racing for :7878, both writing the same
# sd:/slight/ diag files. It presents as a hard 60->30 fps drop on entering training mode, which
# looks like a performance regression in whatever was last changed and is not one. It also
# invalidates every A/B test run while it is there, which is how it cost ~6 rounds once.
#
# Matched on content, not on filename, because the name is exactly what varies -- `.bak`, `.old`,
# `lib_effect_viewer (1).nro`, a hand-renamed known-good build. It is the path constant from the
# plugin's diag module, which every build compiles into rodata; `tests/deploy_plugin.rs` pins that
# the string is still in the source it comes from.
#
# It also matches upstream Jorge SLight builds, and that is deliberate rather than a false
# positive: they bind the same port and write the same files, so having one installed alongside
# this plugin is the same conflict under a different name.
PLUGIN_MARKER = b"sd:/slight/diag.txt"


def is_plugin_copy(path: Path) -> bool:
    """True when `path` is some build of this plugin, whatever it has been renamed to."""
    try:
        return PLUGIN_MARKER in path.read_bytes()
    except OSError:
        return False


def scan_plugins_dir(plugins: Path) -> tuple[list[Path], list[Path]]:
    """Split the files in `plugins` into (strays, unrecognised).

    A stray is a second copy of this plugin under any name. Unrecognised files are reported but
    never block a deploy -- an unrelated Skyline plugin living in the same directory is a normal,
    supported thing to have, and refusing to deploy over one would be wrong.
    """
    known = {INSTALLED_NAME, LEGACY_NAME, *RUNTIME_DEPENDENCIES}
    strays: list[Path] = []
    unrecognised: list[Path] = []
    for entry in sorted(plugins.iterdir()):
        if not entry.is_file() or entry.name in known:
            continue
        (strays if is_plugin_copy(entry) else unrecognised).append(entry)
    return strays, unrecognised


def default_nro() -> Path:
    plugin_root = Path(__file__).resolve().parents[1]
    target_value = os.environ.get("CARGO_TARGET_DIR")
    target = Path(target_value) if target_value else plugin_root / "target"
    if not target.is_absolute():
        target = plugin_root / target
    return target / "aarch64-skyline-switch" / "release" / "lib_effect_viewer.nro"


def copy_missing(source: Path | None, destination: Path, names: tuple[str, ...]) -> None:
    destination.mkdir(parents=True, exist_ok=True)
    for name in names:
        target = destination / name
        candidate = source / name if source is not None else None
        if candidate is not None and candidate.is_file():
            shutil.copy2(candidate, target)
        elif not target.is_file():
            print(f"  !! MISSING dependency: {target}")


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--emulator", choices=("eden", "ryujinx"), required=True)
    parser.add_argument(
        "--mod-dir",
        type=Path,
        help=(
            "Arcropolis mod root; overrides VISIONARY_EDEN_MOD_DIR or "
            "VISIONARY_RYUJINX_MOD_DIR"
        ),
    )
    parser.add_argument(
        "--source-mod-dir",
        type=Path,
        help="optional existing mod root from which Ryujinx dependencies are copied",
    )
    parser.add_argument("--nro", type=Path, default=default_nro())
    parser.add_argument(
        "--remove-strays",
        action="store_true",
        help="delete other copies of this plugin found in the target directory",
    )
    args = parser.parse_args()

    mod_dir = args.mod_dir or (
        eden_mod_directory() if args.emulator == "eden" else ryujinx_mod_directory()
    )
    mod_dir = mod_dir.expanduser().resolve()
    nro = args.nro.expanduser().resolve()
    if not nro.is_file():
        raise SystemExit(f"Plugin not found at {nro}; build it first")

    source_mod = args.source_mod_dir
    if source_mod is None and args.emulator == "ryujinx":
        source_mod = eden_mod_directory()
    if source_mod is not None:
        source_mod = source_mod.expanduser().resolve()

    plugins = mod_dir / "romfs" / "skyline" / "plugins"
    plugins.mkdir(parents=True, exist_ok=True)
    if args.emulator == "ryujinx":
        source_exefs = source_mod / "exefs" if source_mod and source_mod.is_dir() else None
        copy_missing(source_exefs, mod_dir / "exefs", SKYLINE_EXEFS)
        source_plugins = (
            source_mod / "romfs" / "skyline" / "plugins"
            if source_mod and source_mod.is_dir()
            else None
        )
        copy_missing(source_plugins, plugins, RUNTIME_DEPENDENCIES)

    legacy = plugins / LEGACY_NAME
    if legacy.is_file():
        legacy.unlink()

    strays, unrecognised = scan_plugins_dir(plugins)
    for entry in unrecognised:
        print(f"  -- not one of ours, left alone: {entry.name}")
    if strays and not args.remove_strays:
        # Refuse BEFORE copying, so a refusal changes nothing and can be re-run after the user
        # has looked at the files themselves. Deleting anything in someone's mod directory
        # without being asked is not this script's call to make -- one of these is quite
        # plausibly a known-good build somebody parked there on purpose.
        print()
        print("REFUSING TO DEPLOY: a second copy of this plugin is already installed.")
        for entry in strays:
            print(f"  {entry}")
        print()
        print(
            "Skyline loads every file in this directory as a plugin, extension ignored, so each\n"
            "of the above runs a second full copy: double ACMD hooks, double per-frame ticks,\n"
            "two servers fighting over :7878. It shows up as a 60->30 fps drop in training mode\n"
            "and it invalidates any A/B test run while it is there."
        )
        print()
        print("Move them somewhere outside this directory, or re-run with --remove-strays.")
        raise SystemExit(2)
    for entry in strays:
        entry.unlink()
        print(f"  Removed second plugin copy: {entry}")

    installed = plugins / INSTALLED_NAME
    shutil.copy2(nro, installed)
    print(f"Deployed {installed}")

    missing = [name for name in RUNTIME_DEPENDENCIES if not (plugins / name).is_file()]
    for name in missing:
        print(f"  !! MISSING dependency: {plugins / name}")
    if not missing:
        print("All runtime plugin dependencies are present.")
        smashline = plugins / "libsmashline_plugin.nro"
        if b"smashline_install_state_callback" not in smashline.read_bytes():
            print(
                "  !! WARNING: libsmashline_plugin.nro may be Smashline 1; "
                "the Smashline 2 callback export was not found."
            )

    if args.emulator == "ryujinx":
        print(f"Runtime diagnostics: {ryujinx_sd_directory() / 'slight' / 'diag.txt'}")
    print(f"Restart {args.emulator.title()} fully so the new plugin loads.")


if __name__ == "__main__":
    main()

File: tests/deploy_plugin.rs
//! The plugin deploy script's double-plugin guard, exercised end to end.
//!
//! Skyline loads *every* file in `romfs:/skyline/plugins/` as a plugin, extension ignored, so a
//! `lib_effect_viewer.nro.bak` left beside the real one runs a second full copy of the plugin:
//! double ACMD hooks, double per-frame drivers, two servers racing for :7878. The visible symptom
//! is a hard 60→30 fps drop entering training mode, which reads as a performance regression in
//! whatever was last touched. It cost about six rounds of misdiagnosis once, because it also
//! invalidates every A/B test run while it is there.
//!
//! `deploy_plugin.py` now refuses to deploy over one. This is where that is checked, rather than
//! in the Python file itself, because `cargo test` is the gate this project actually runs — a
//! test suite nothing invokes is a comment.
//!
//! The tests drive the real script in a throwaway directory. Nothing here mocks the filesystem,
//! so what passes is what a user's mod folder will do.

use std::fs;
use std::path::{Path, PathBuf};
use std::process::{Command, Output};

/// The path constant the script searches candidate files for. Copied here on purpose: the script
/// hardcodes these bytes, and [`the_marker_the_script_greps_for_is_still_in_the_plugin_source`]
/// checks all three copies still agree.
const PLUGIN_MARKER: &str = "sd:/slight/diag.txt";

fn repo_root() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
}

fn script() -> PathBuf {
    repo_root().join("plugins/slight_replica/tools/deploy_plugin.py")
}

/// A fresh, empty mod root under the system temp directory, plus its plugins directory. Each
/// test owns a distinct name so the suite is safe to run in parallel.
fn fixture(name: &str) -> (PathBuf, PathBuf) {
    let root = std::env::temp_dir().join(format!("visionary-deploy-test/{name}"));
    let _ = fs::remove_dir_all(&root);
    let plugins = root.join("romfs/skyline/plugins");
    fs::create_dir_all(&plugins).expect("create fixture plugins dir");
    (root, plugins)
}

/// Bytes that look like a build of this plugin to the script: anything containing the marker.
fn plugin_bytes(tag: &str) -> Vec<u8> {
    format!("\0ELF-ish padding\0{PLUGIN_MARKER}\0build={tag}\0").into_bytes()
}

fn write(path: &Path, bytes: &[u8]) {
    fs::write(path, bytes).unwrap_or_else(|e| panic!("write {}: {e}", path.display()));
}

/// Lay down the three runtime dependencies the script expects, so its "MISSING dependency"
/// reporting is not what a test ends up measuring.
fn write_dependencies(plugins: &Path) {
    write(&plugins.join("libarcropolis.nro"), b"arcropolis");
    write(&plugins.join("libnro_hook.nro"), b"nro_hook");
    write(
        &plugins.join("libsmashline_plugin.nro"),
        b"smashline_install_state_callback",
    );
}

fn run_deploy(mod_dir: &Path, nro: &Path, extra: &[&str]) -> Output {
    let mut cmd = Command::new("python3");
    cmd.arg(script())
        .arg("--emulator")
        .arg("eden")
        .arg("--mod-dir")
        .arg(mod_dir)
        .arg("--nro")
        .arg(nro)
        .args(extra);
    cmd.output().expect(
        "run deploy_plugin.py — python3 is a hard dependency of the plugin build and deploy \
         scripts (scripts/build.sh execs it), so a missing interpreter is a real failure here, \
         not a reason to skip",
    )
}

fn combined(out: &Output) -> String {
    format!(
        "{}{}",
        String::from_utf8_lossy(&out.stdout),
        String::from_utf8_lossy(&out.stderr)
    )
}

/// The paired positive for the two "did not deploy" assertions below.
///
/// Without this, "the installed .nro is absent" would pass for a fixture that could never have
/// produced one — a broken script path, a Python syntax error, a wrong argument name all look
/// exactly like a successful refusal.
#[test]
fn a_clean_plugins_directory_deploys() {
    let (root, plugins) = fixture("clean");
    write_dependencies(&plugins);
    let nro = root.join("lib_effect_viewer.nro");
    write(&nro, &plugin_bytes("new"));

    let out = run_deploy(&root, &nro, &[]);
    assert!(
        out.status.success(),
        "expected success, got:\n{}",
        combined(&out)
    );
    assert_eq!(
        fs::read(plugins.join("lib_effect_viewer.nro")).unwrap(),
        plugin_bytes("new"),
        "a clean directory must actually receive the build"
    );
}

#[test]
fn a_stray_copy_of_the_plugin_refuses_the_deploy_and_changes_nothing() {
    let (root, plugins) = fixture("stray");
    write_dependencies(&plugins);
    let stray = plugins.join("lib_effect_viewer.nro.bak");
    write(&stray, &plugin_bytes("old"));
    let nro = root.join("lib_effect_viewer.nro");
    write(&nro, &plugin_bytes("new"));

    let out = run_deploy(&root, &nro, &[]);
    let text = combined(&out);
    assert!(!out.status.success(), "expected a refusal, got:\n{text}");
    assert!(
        text.contains("lib_effect_viewer.nro.bak"),
        "the refusal must name the file to remove, or it is another mystery:\n{text}"
    );
    assert!(
        !plugins.join("lib_effect_viewer.nro").exists(),
        "a refusal must be a no-op — it is safe to re-run only if it deployed nothing"
    );
    assert_eq!(
        fs::read(&stray).unwrap(),
        plugin_bytes("old"),
        "and it must not have deleted the stray it refused over"
    );
}

/// The name is the one thing that varies. `.bak`, `.old`, a hand-renamed known-good build — the
/// script matches on content for exactly this reason, so the test refuses to only try `.bak`.
#[test]
fn a_stray_is_found_under_any_name() {
    for name in ["lib_effect_viewer.old", "keep-this-one.nro", "backup.txt"] {
        let (root, plugins) = fixture("renamed");
        write_dependencies(&plugins);
        write(&plugins.join(name), &plugin_bytes("old"));
        let nro = root.join("lib_effect_viewer.nro");
        write(&nro, &plugin_bytes("new"));

        let out = run_deploy(&root, &nro, &[]);
        let text = combined(&out);
        assert!(
            !out.status.success(),
            "{name} should have been caught:\n{text}"
        );
        assert!(text.contains(name), "{name} should be named in:\n{text}");
    }
}

/// An unrelated Skyline plugin in the same directory is normal and supported. Refusing to deploy
/// over one would make the guard something users route around, at which point it protects nobody.
#[test]
fn an_unrelated_plugin_does_not_block_the_deploy() {
    let (root, plugins) = fixture("unrelated");
    write_dependencies(&plugins);
    write(
        &plugins.join("libtraining_modpack.nro"),
        b"some other plugin",
    );
    let nro = root.join("lib_effect_viewer.nro");
    write(&nro, &plugin_bytes("new"));

    let out = run_deploy(&root, &nro, &[]);
    let text = combined(&out);
    assert!(out.status.success(), "expected success, got:\n{text}");
    assert!(
        plugins.join("libtraining_modpack.nro").exists(),
        "and it must still be there afterwards"
    );
}

#[test]
fn remove_strays_deletes_the_second_copy_and_then_deploys() {
    let (root, plugins) = fixture("remove");
    write_dependencies(&plugins);
    let stray = plugins.join("lib_effect_viewer.nro.bak");
    write(&stray, &plugin_bytes("old"));
    let nro = root.join("lib_effect_viewer.nro");
    write(&nro, &plugin_bytes("new"));

    let out = run_deploy(&root, &nro, &["--remove-strays"]);
    let text = combined(&out);
    assert!(out.status.success(), "expected success, got:\n{text}");
    assert!(!stray.exists(), "the stray should be gone:\n{text}");
    assert_eq!(
        fs::read(plugins.join("lib_effect_viewer.nro")).unwrap(),
        plugin_bytes("new")
    );
}

/// The guard is only as good as the marker, and the marker is a string literal duplicated in three
/// places that no compiler relates to each other.
///
/// If the plugin ever stops writing `sd:/slight/diag.txt` — renamed file, path built by `format!`
/// instead of a constant — every test above still passes, because they synthesise their own
/// fixture bytes containing the marker. The scan would then match nothing real and the guard would
/// go quietly dead. This is the assertion that fails instead.
///
/// **What it does not prove.** That the string survives into the linked `.nro`. It does today
/// (verified by grepping a release build), and it is used as a `&str` constant in two `std::fs`
/// calls so the linker has no reason to drop it — but a built plugin does not exist in a fresh
/// clone, and a check that skips when the artefact is missing would pass vacuously in exactly the
/// situation where it matters. Source is the strongest thing available unconditionally.
#[test]
fn the_marker_the_script_greps_for_is_still_in_the_plugin_source() {
    let diag = repo_root().join("plugins/slight_replica/src/slight/diag.rs");
    let source = fs::read_to_string(&diag).expect("read the plugin's diag module");
    assert!(
        source.contains(&format!("\"{PLUGIN_MARKER}\"")),
        "{} no longer contains the literal \"{PLUGIN_MARKER}\" that deploy_plugin.py scans \
         candidate files for. Either restore it or change PLUGIN_MARKER in the script, this \
         test, and the plugin together — the guard silently matches nothing otherwise.",
        diag.display()
    );

    let script_source = fs::read_to_string(script()).expect("read deploy_plugin.py");
    assert!(
        script_source.contains(&format!("b\"{PLUGIN_MARKER}\"")),
        "deploy_plugin.py's PLUGIN_MARKER has drifted from this test's copy of it"
    );
}
